File: examples/python/index/quick-example.py
from aws_durable_execution_sdk_python import (
    DurableContext,
    durable_execution,
    durable_step,
)
from aws_durable_execution_sdk_python.config import Duration


@durable_step
def validate_order(order_id: str) -> dict:
    # Validation logic here
    return {"order_id": order_id, "valid": True}


@durable_step
def charge_payment(order_id: str, amount: float) -> dict:
    # Payment processing logic here
    return {"transaction_id": "txn_123", "status": "completed"}


@durable_step
def fulfill_order(order_id: str) -> dict:
    # Fulfillment logic here
    return {"tracking_number": "TRK123456"}


@durable_execution
def process_order(event: dict, context: DurableContext) -> dict:
    order_id = event["order_id"]
    amount = event["amount"]

    # Step 1: Validate the order
    validation = context.step(validate_order(order_id))

    if not validation["valid"]:
        return {"status": "failed", "reason": "invalid_order"}

    # Step 2: Charge payment
    payment = context.step(charge_payment(order_id, amount))

    # Step 3: Wait for payment confirmation (simulated)
    context.wait(Duration.from_seconds(5))

    # Step 4: Fulfill the order
    fulfillment = context.step(fulfill_order(order_id))

    return {
        "status": "completed",
        "order_id": order_id,
        "transaction_id": payment["transaction_id"],
        "tracking_number": fulfillment["tracking_number"],
    }

File: examples/java/index/quick-example.java
public class ProcessOrderHandler extends DurableHandler<Map<String, Object>, Map<String, Object>> {

    @Override
    public Map<String, Object> handleRequest(Map<String, Object> event, DurableContext ctx) {
        var orderId = (String) event.get("order_id");
        var amount = (Double) event.get("amount");

        // Step 1: Validate the order
        var validation = ctx.step("validate-order", Map.class, stepCtx -> {
            // Validation logic here
            return Map.of("order_id", orderId, "valid", true);
        });

        if (!(boolean) validation.get("valid")) {
            return Map.of("status", "failed", "reason", "invalid_order");
        }

        // Step 2: Charge payment
        var payment = ctx.step("charge-payment", Map.class, stepCtx -> {
            // Payment processing logic here
            return Map.of("transaction_id", "txn_123", "status", "completed");
        });

        // Step 3: Wait for payment confirmation (simulated)
        ctx.wait("payment-confirmation", Duration.ofSeconds(5));

        // Step 4: Fulfill the order
        var fulfillment = ctx.step("fulfill-order", Map.class, stepCtx -> {
            // Fulfillment logic here
            return Map.of("tracking_number", "TRK123456");
        });

        return Map.of(
            "status", "completed",
            "order_id", orderId,
            "transaction_id", payment.get("transaction_id"),
            "tracking_number", fulfillment.get("tracking_number")
        );
    }
}
